src/System.Buffers/src/System/Buffers/NativeBufferPool.cs

```diff
@@ -8,12 +8,31 @@
 
 namespace System.Buffers
 {
-    public unsafe sealed class NativeBufferPool<T> where T : struct
+    public unsafe sealed class NativeBufferPool<T> : IDisposable where T : struct
     {
         private const int NUMBER_OF_BUFFERS_IN_BUCKET = 50;
         private int _maxElements;
         private NativeBufferBucket<T>[] _buckets;
 
+        private static NativeBufferPool<byte> _sharedInstance = null;
+
+        public static NativeBufferPool<byte> SharedByteBufferPool
+        {
```

Comment:
This algorithm works as follows:
1) Check if the shared instance is null; if it has a value, simply return it
2) If it is null, allocate a new one
3) Do an Interlocked.CompareExchange on the shared instance; if there was another thread that beat us to the CompareExchange (meaning by the time we got here, another thread already did the same thing but quicker) then dispose of the instance we created since it won't be used
4) Return the shared instance, which is guaranteed to be non-null now

A mutex is a very heavy hammer to solve this problem; Interlocked and Volatile are much lighter weight to achieve the same result, meaning we have a much smaller perf hit. This class is also trying to be as lock-free as possible